text/0000-bee-model-crate/0000-bee-model-crate.md

```diff
@@ -0,0 +1,277 @@
+| `attachment_timestamp_lowerbound` | *not specified*                                        | 27              |
+| `attachment_timestamp_upperbound` | *not specified*                                        | 27              |
+| `nonce`                           | the Proof-of-Work nonce of the transaction             | 81              |
+
+Each transaction is uniquely identified by its *transaction hash*, which is
+calculated based on all fields of the transaction. Note that the transaction
+hash is not part of the transaction.
+
+Because modifying the fields of a transaction would invalidate its hash,
+`Transaction` is immutable after creation.
+
+## Transaction struct
+
+A transaction is represented by the `Transaction` struct and follows the structure presented
+in the table above:
+
+```rust
+pub struct Transaction {
+    signature_or_message_fragment: SignatureOrMessageFragment,
+    address: Address,
+    value: Value,
+    obsolete_tag: Tag,
+    timestamp: Timestamp,
+    current_index: Index,
+    last_index: Index,
+    bundle_hash: BundleHash,
+    trunk: TransactionHash,
+    branch: TransactionHash,
+    tag: Tag,
+    attachment_timestamp: Timestamp,
+    attachment_timestamp_lower_bound: Timestamp,
+    attachment_timestamp_upper_bound: Timestamp,
+    nonce: Nonce
+}
+```
+
+We treat the types of the `Transaction` struct's fields as opaque newtypes for
+now so that we can flesh them out during the implementation phase or in future
+RFCs.
+
+```rust
+pub struct Trit(u8);
+pub struct SignatureOrMessageFragment([Trit; 6561]);
+pub struct Address([Trit; 243]);
+pub struct Tag([Trit; 81]);
+pub struct Timestamp(u64);
+pub struct Index(u64);
+pub struct Value(i64);
+pub struct BundleHash([Trit; 243]);
+pub struct TransactionHash([Trit; 243]);
+pub struct Nonce([Trit; 81]);
+```
+
+The `Transaction` type only contains two constructor methods getter methods for
+read-only borrows of its fields.
+
+```rust
+impl Transaction {
+    /// Create a `Transaction` from a reader.
+    pub fn from_reader<R: std::io::Read>(reader: R) -> Result<Self, TransactionError> {
+        unimplemented!()
+    }
+
+    /// Create a `Transaction` from a slice of bytes
+    pub fn from_slice(stream: &[u8]) -> Result<Self, TransactionError> {
+        unimplemented!()
+    }
+
+    pub fn signature_or_message_fragment(&self) -> &SignatureOrMessageFragment {
+        &self.signature_fragments
+    }
+
+    pub fn address(&self) -> &Address {
+        &self.address
+    }
+
+    pub fn value(&self) -> &Value {
+        &self.value
+    }
+
+    ...
+}
+```
+
+The `from_slice` and `from_reader` methods will use a `TransactionDraft` internally,
+but these are implementation details.
+
+The `TransactionError` is not fully fleshed out yet. For the time being, it
+definitely contains `io::Error`:
+
+```rust
+pub enum TransactionError {
+    Io(io::Error),
+}
```

Comment:
I think we should also start considering which error handling library to use. I saw there's another new one called [thiserror](https://github.com/dtolnay/thiserror). What do you think?